prow/plugins/buildifier/buildifier.go

```diff
@@ -0,0 +1,218 @@
+// WORKSPACE, and skylark (.bzl) files in pull requests.
+package buildifier
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/bazelbuild/buildtools/build"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/genfiles"
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName  = "buildifier"
+	maxComments = 20
+)
+
+var buildifyRe = regexp.MustCompile(`(?mi)^/buildif(y|ier)\s*$`)
+
+func init() {
+	plugins.RegisterGenericCommentHandler(pluginName, handleGenericComment)
+}
+
+type githubClient interface {
+	GetFile(org, repo, filepath, commit string) ([]byte, error)
+	GetPullRequest(org, repo string, number int) (*github.PullRequest, error)
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	CreateReview(org, repo string, number int, r github.DraftReview) error
+	ListPullRequestComments(org, repo string, number int) ([]github.ReviewComment, error)
+}
+
+func handleGenericComment(pc plugins.PluginClient, e github.GenericCommentEvent) error {
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, &e)
+}
+
+// modifiedBazelFiles returns a map from filename to patch string for all Bazel files
+// that are modified in the PR.
+func modifiedBazelFiles(ghc githubClient, org, repo string, number int, sha string) (map[string]string, error) {
+	changes, err := ghc.GetPullRequestChanges(org, repo, number)
+	if err != nil {
+		return nil, err
+	}
+
+	gfg, err := genfiles.NewGroup(ghc, org, repo, sha)
+	if err != nil {
+		return nil, err
+	}
+
+	modifiedFiles := make(map[string]string)
+	for _, change := range changes {
+		switch {
+		case gfg.Match(change.Filename):
+			continue
+		case change.Status == github.PullRequestFileRemoved || change.Status == github.PullRequestFileRenamed:
+			continue
+		case strings.Contains(change.Filename, "BUILD"):
+			break
+		case strings.Contains(change.Filename, "WORKSPACE"):
+			break
+		case filepath.Ext(change.Filename) != ".bzl":
+			continue
+		}
+		modifiedFiles[change.Filename] = change.Patch
+	}
+	return modifiedFiles, nil
+}
+
+func uniqProblems(problems []string) []string {
+	sort.Strings(problems)
+	var uniq []string
+	last := ""
+	for _, s := range problems {
+		if s != last {
+			last = s
+			uniq = append(uniq, s)
+		}
+	}
+	return uniq
+}
+
+// problemsInFiles runs buildifier on the files. It returns a map from the file to
+// a list of problems with that file.
+func problemsInFiles(r *git.Repo, files map[string]string) (map[string][]string, error) {
+	problems := make(map[string][]string)
+	for f := range files {
+		src, err := ioutil.ReadFile(filepath.Join(r.Dir, f))
+		if err != nil {
+			return nil, err
+		}
+		// This is modeled after the logic from buildifier:
+		// https://github.com/bazelbuild/buildtools/blob/master/buildifier/buildifier.go#L261
```

Comment:
These links tend to break over time. If you want it to persist, you should use a commit hash.

https://help.github.com/articles/getting-permanent-links-to-files/